### PR TITLE
Kubeadm phases - improve flags management

### LINKED ISSUE #1206: Improve flag management for kubedm phases
as per comment https://github.com/kubernetes/kubernetes/pull/70405#discussion_r229397171

Currently flags management for phases is very simple: all autogenerated command for phases will get the same flags of the parent phase.

It is necessary to improve flags management supporting:
- the option to add flags that applies only to autogenerated command for phases (not to the parent command)
- the option to select the subset of relevant flags for each autogenerated command

@neolit123 
@yagonobre

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR makes flags for cmd automatically generated for phases equal to the hardcoded flags before refactor. 

**Which issue(s) this PR fixes** : 

Fixes: https://github.com/kubernetes/kubeadm/issues/1206

**Special notes for your reviewer**:

The solution guarantees that flags will be always defined once, in the parent command.

Autogenerated commands get flags according to this rule:

- global flags will be always inherited by autogenerated commands (this is managed by cobra)
- local flags might be inherited by autogenerated commands, but this requires explicit opt-in so each phase can select the subset of relevant flags 
- it is possible to define additional flags that might be inherited by autogenerated commands via explicit opt-in, but are not applied to the parent command

Phases are updated accordingly

/sig cluster-lifecycle
/kind feature
/assign @timothysc
/assign @neolit123 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@yagonobre

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -365,7 +385,12 @@ func (d initData) KubeConfigDir() string {
 	if d.dryRun {
 		return d.dryRunDir
 	}
-	return kubeadmconstants.KubernetesDir
+	return d.kubeconfigDir
+}
+
+// KubeConfigPath returns the path to the kubeconfig file to use for connecting to kubernetes
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: nit: Kubernetes uppercase.

**Review by neolit123 — APPROVED** [review]
@fabriziopandini thanks, LGTM.
i like that the flags are defined in init.go and only marked for usage in sub-phases.
i think the solution is good.

File: cmd/kubeadm/app/cmd/phases/etcd.go [hunk 2]
```diff
@@ -53,20 +53,30 @@ func NewEtcdPhase() workflow.Phase {
...
+		Name:     "local",
+		Short:    "Generates the static Pod manifest file for a local, single-node local etcd instance.",
+		Example:  etcdLocalExample,
+		Run:      runEtcdPhaseLocal(),
+		CmdFlags: getEtcdPhaseFlags(),
 	}
 	return phase
 }
 
+func getEtcdPhaseFlags() []string {
```
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/etcd.go]: Maybe we can define this flags on `newEtcdLocalSubPhase`

**Review by yagonobre — APPROVED** [review]
Good work! LGTM

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 3]
```diff
@@ -105,11 +106,28 @@ func newCertSubPhase(certSpec *certsphase.KubeadmCert, run func(c workflow.RunDa
...
-		Run: run,
+		Run:      run,
+		CmdFlags: getCertPhaseFlags(certSpec.Name),
 	}
 	return phase
 }
 
+func getCertPhaseFlags(name string) []string {
+	flags := []string{
+		"config",
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/certs.go]: This should all be const-i-fied in some central place with a comment block outlining the order.

**Review by timothysc — CHANGES_REQUESTED** [review]
Minor comment that percolates to all the sites. 
/approve 
of the effort, but without a central source of truth it's pretty tough to reason about.

File: cmd/kubeadm/app/cmd/init.go [hunk 4]
```diff
@@ -155,11 +160,22 @@ func NewCmdInit(out io.Writer) *cobra.Command {
...
+	// init command local flags could be eventually inherited by the sub-commands automatically generated for phases
+	AddInitConfigFlags(cmd.Flags(), options.externalcfg, &options.featureGatesString)
+	AddInitOtherFlags(cmd.Flags(), &options.cfgPath, &options.skipTokenPrint, &options.dryRun, &options.ignorePreflightErrors)
+	options.bto.AddTokenFlag(cmd.Flags())
+	options.bto.AddTTLFlag(cmd.Flags())
+
+	// defines additional flag that are not used by the init command but that could be eventually used
+	// by the sub-commands automatically generated for phases
+	initRunner.SetPhaseSubcommandsAdditionalFlags(func(flags *flag.FlagSet) {
+		flags.Var(utilflag.NewMapStringString(&options.externalcfg.APIServerExtraArgs), "apiserver-extra-args", "A set of extra flags to pass to the API Server or override default ones in form of <flagname>=<value>")
```
Comment (liztio) [inline cmd/kubeadm/app/cmd/init.go]: could we move all these flags into functions? some already exist at https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/options/generic.go and it'd be best to have the prose in one place

File: cmd/kubeadm/app/cmd/phases/etcd.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/etcd.go]: this method is used both in `newEtcdLocalSubPhase` and in `NewEtcdPhase`.

**Comment by fabriziopandini** [comment]
@timothysc @neolit123 @yagonobre @liztio thanks for the feedback!
Everything should be addressed now

@timothysc I applied const only to things in the scope of this PR. After our code audit, we will eventually file an issue for extending the same approach to the other kubeadm commands.

**Comment by neolit123** [comment]
/retest
/lgtm

LGTM-ing here, had another pass at the whole diff, new constants look good.

**Pull request merged**